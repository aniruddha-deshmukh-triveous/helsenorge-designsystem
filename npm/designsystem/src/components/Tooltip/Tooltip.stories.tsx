import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';

import Tooltip from './Tooltip';
import TooltipExample from '../TooltipExample';

export default {
  title: 'Components/Tooltip',
  component: Tooltip,
  argTypes: {
    children: {
      control: 'text',
      defaultValue: 'ordet',
    },
    description: {
      control: 'text',
      defaultValue: 'Dette er tekst som skal fylle HelpBubble',
    },
  },
} as ComponentMeta<typeof Tooltip>;

export const Default: ComponentStory<typeof Tooltip> = (args: any) => (
  <div>
    <TooltipExample {...args} />
  </div>
);
