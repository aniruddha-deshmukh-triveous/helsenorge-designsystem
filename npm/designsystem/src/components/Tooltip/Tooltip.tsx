import React, { useContext, useState, useEffect, useRef, ReactChild, ReactChildren, ReactNode } from 'react';
import { AnalyticsId } from '../../constants';
import { useUuid } from '../../hooks/useUuid';

import HelpBubble from '../HelpBubble';

import tooltipstyles from './styles.module.scss';

export interface TooltipProps {
  /**Ordet som skal ha en tilh√∏rende hjelpeblubb */
  children: ReactNode;
  /**Teksten som skal vises i hjelpeblubben */
  description: ReactNode;
  /**Valgfri test-id */
  testId?: string;
}

export const Tooltip: React.FC<TooltipProps> = (props: TooltipProps): JSX.Element => {
  const { children, description, testId } = props;

  const HOVER_DELAY = 200;

  const ariaDescribedBy = `help-bubble-${useUuid()}`;
  const node = useRef<HTMLDivElement>(null);
  const firstRender = useRef(true);
  const anyTooltipIsOpen = useTooltipOpen();
  const toggleTooltipIsOpen = useTooltipOpenToggle();
  const [visTooltip, setVisTooltip] = useState(false);
  const [hoverVisTooltip, setHoverVisTooltip] = useState(false);
  const [delayHandler, setDelayHandler] = useState<number | null>(null);

  useEffect(() => {
    if (!firstRender.current) toggleTooltipIsOpen();
    if (firstRender.current) firstRender.current = false;
  }, [visTooltip]);

  const handleClick = (): void => {
    if (!hoverVisTooltip) {
      setVisTooltip(false);
    }
  };

  const handleMouseEnter = (): void => {
    if (!anyTooltipIsOpen) {
      setDelayHandler(
        window.setTimeout(() => {
          setHoverVisTooltip(true);
        }, HOVER_DELAY)
      );
    }
  };

  const handleMouseLeave = (): void => {
    setHoverVisTooltip(false);
    delayHandler && clearTimeout(delayHandler);
  };

  useEffect(() => {
    document.addEventListener('mouseup', handleClick);
    return (): void => {
      document.removeEventListener('mouseup', handleClick);
      delayHandler && clearTimeout(delayHandler);
    };
  }, []);

  const handleKeyDown = (e: React.KeyboardEvent<HTMLDivElement>): void => {
    if (e.key === 'Enter') {
      hoverVisTooltip ? setVisTooltip(false) : setVisTooltip(!visTooltip);
      setHoverVisTooltip(false);
    }
    if (e.key === 'Escape') {
      setVisTooltip(false);
      setHoverVisTooltip(false);
    }
  };

  const handleTooltipClick = () => {
    setVisTooltip(!visTooltip);
    setHoverVisTooltip(false);
  };

  return (
    <>
      <span
        className={tooltipstyles.word}
        ref={node}
        tabIndex={0}
        onClick={(): void => handleTooltipClick()}
        onMouseEnter={(): void => handleMouseEnter()}
        onMouseLeave={(): void => handleMouseLeave()}
        onFocus={(): void => handleMouseEnter()}
        onBlur={(): void => handleMouseLeave()}
        onKeyDown={handleKeyDown}
        aria-describedby={ariaDescribedBy}
        data-testid={testId}
        data-analyticsid={AnalyticsId.Tooltip}
      >
        {children}
      </span>
      <HelpBubble helpBubbleId={ariaDescribedBy} controllerRef={node} showBubble={visTooltip || hoverVisTooltip} noCloseButton>
        {description}
      </HelpBubble>
    </>
  );
};

const TooltipOpenContext = React.createContext(false);
const TooltipOpenToggleContext = React.createContext(() => {});

interface ContextProps {
  children: ReactChild | ReactChildren;
}
export const useTooltipOpen = () => {
  return useContext(TooltipOpenContext);
};
export const useTooltipOpenToggle = () => {
  return useContext(TooltipOpenToggleContext);
};

export const TooltipOpenProvider = ({ children }: ContextProps) => {
  const [isOpen, setIsOpen] = useState(false);

  const tooltipIsOpen = () => {
    setIsOpen(prevIsOpen => !prevIsOpen);
  };

  return (
    <TooltipOpenContext.Provider value={isOpen}>
      <TooltipOpenToggleContext.Provider value={tooltipIsOpen}>{children}</TooltipOpenToggleContext.Provider>
    </TooltipOpenContext.Provider>
  );
};

export default Tooltip;
