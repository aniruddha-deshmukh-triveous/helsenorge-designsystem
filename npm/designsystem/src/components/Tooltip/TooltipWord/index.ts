import TooltipWord from './TooltipWord';
export * from './TooltipWord';
export default TooltipWord;
