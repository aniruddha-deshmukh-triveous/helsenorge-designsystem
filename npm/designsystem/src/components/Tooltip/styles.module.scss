@import '../../scss/_palette.scss';
@import '../../scss/_spacers.scss';

.word {
  position: relative;
  display: inline-block;
  -webkit-text-decoration: underline dashed $plum400;
  text-decoration: underline 0.15rem dashed $plum400;

  &:hover {
    cursor: help;
  }

  &:focus-visible {
    box-shadow: 0 0 0 getSpacer(4xs) $black;
    outline: none;
  }
}
