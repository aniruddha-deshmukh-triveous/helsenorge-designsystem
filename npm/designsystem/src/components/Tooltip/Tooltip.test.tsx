import * as React from 'react';

import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import Tooltip from './Tooltip';

describe('Gitt at Tooltip rendres ', () => {
  describe('Når komponenten får en tekst', () => {
    it('Så rendres den riktig', () => {
      const { container } = render(<Tooltip description={'beskrivelse av ordet'}>{'ordet'}</Tooltip>);

      const word = screen.getByText('ordet');

      expect(word).toBeInTheDocument();
      expect(word).toHaveClass('word');

      expect(container).toMatchSnapshot();
    });
  });

  describe('Når tekst trykkes på', () => {
    it('Så vises HelpBubble', async () => {
      render(<Tooltip description={'beskrivelse av ordet'}>{'ordet'}</Tooltip>);

      const word = screen.getByText('ordet');
      await userEvent.click(word);

      const helpBubble = screen.getByRole('tooltip', { name: 'beskrivelse av ordet' });
      expect(helpBubble).toBeInTheDocument();
    });
  });

  describe('Når tekst hovres med musen', () => {
    it('Så vises HelpBubble', async () => {
      render(<Tooltip description={'beskrivelse av ordet'}>{'ordet'}</Tooltip>);

      const word = screen.getByText('ordet');
      await userEvent.hover(word);

      const helpBubble = screen.getByRole('tooltip', { name: 'beskrivelse av ordet' });
      expect(helpBubble).toBeInTheDocument();
    });
  });

  describe('Når tekst får fokus', () => {
    it('Så vises HelpBubble', async () => {
      render(<Tooltip description={'beskrivelse av ordet'}>{'ordet'}</Tooltip>);

      const word = screen.getByText('ordet');

      await userEvent.tab();
      expect(word).toHaveFocus();

      const helpBubble = screen.getByRole('tooltip', { name: 'beskrivelse av ordet' });
      expect(helpBubble).toBeInTheDocument();
    });
  });

  describe('Når det trykkes mens HelpBubble er åpen', () => {
    it('Så skjules HelpBubble', async () => {
      render(<Tooltip description={'beskrivelse av ordet'}>{'ordet'}</Tooltip>);

      const word = screen.getByText('ordet');
      await userEvent.click(word);

      const helpBubble = screen.getByRole('tooltip');
      await userEvent.click(helpBubble);

      expect(helpBubble).not.toHaveClass('helpbubble--visible');
    });
  });
});
