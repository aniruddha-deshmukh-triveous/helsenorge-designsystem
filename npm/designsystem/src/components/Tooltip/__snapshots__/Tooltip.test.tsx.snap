// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at Tooltip rendres  Når komponenten får en tekst Så rendres den riktig 1`] = `
<div>
  <button
    aria-describedby="test"
    class="word"
    data-analyticsid="tooltip"
  >
    ordet
  </button>
</div>
`;
